File: AdventOfCode2022/Python/Day12/Day12.py
altitude = 'abcdefghijklmnopqrstuvwxyz'

def ParseData(file):    
    """
    function to read and parse in data from a txt file
    """
    with open(file) as f:
        data = f.read().split('\n')

    # find starting pos and ending pos
    rows = []
    for y in range(0,len(data)):
        row = []
        for x in range(0,len(data[0])):
            
            if data[y][x] == "S":
                init_pos = [x,y]

            if data[y][x] == "E":
                final_pos = [x,y]
            
            row.append({"altitude":data[y][x], "dist_from_E":''})
        rows.append(row)
    
    rows[init_pos[1]][init_pos[0]] = {"altitude":'a', "dist_from_E":''}
    rows[final_pos[1]][final_pos[0]] = {"altitude":'z', "dist_from_E":0}

    return rows, init_pos, final_pos

def Part1(filename):
    """
    Solutions to Part 1
    """

    rows, init_pos, final_pos = ParseData(filename)
    paths = rows
    i = 0
    
    while type(paths[init_pos[1]][init_pos[0]]["dist_from_E"]) != int:

        for y in range(0,len(rows)):
            for x in range(0,len(rows[0])):

                if paths[y][x]["dist_from_E"] == i:
    
                    current_square = paths[y][x]

                    coord_left  = [x-1,y  ]
                    coord_right = [x+1,y  ]
                    coord_up    = [x  ,y-1]
                    coord_down  = [x  ,y+1]

                    for [x_check,y_check] in [coord_down,coord_up,coord_left,coord_right]:
                    
                        # ignore if out of bounds
                        if len(paths[0]) -1 < x_check or x_check <0 or len(paths) -1 < y_check or y_check < 0 :
                            continue
                        
                        #ignore if already int
                        if type(paths[y_check][x_check]["dist_from_E"]) == int and paths[y_check][x_check]["dist_from_E"] < current_square["dist_from_E"]+1:
                            continue

                        next_square = paths[y_check][x_check]
                        current_altitude = altitude.index(current_square["altitude"])
                        next_altitude    = altitude.index(next_square["altitude"])

                        if next_altitude + 2 > current_altitude :
                            paths[y_check][x_check]["dist_from_E"] = current_square["dist_from_E"]+1                        
        
        i += 1
    
    elements = []
    for row in paths:
        for square in row:
            if type(square["dist_from_E"])== int:
                elements.append(square["dist_from_E"])

    return max(elements)

def Part2(filename):
    """
    Solutions to Part 2
    """
    
    rows, init_pos, final_pos = ParseData(filename)
    paths = rows
    i = 0

    while type(paths[init_pos[1]][init_pos[0]]["dist_from_E"]) != int:

        for y in range(0,len(rows)):
            for x in range(0,len(rows[0])):
                
                if paths[y][x]["dist_from_E"] == i:
    
                    current_square = paths[y][x]

                    coord_left  = [x-1,y  ]
                    coord_right = [x+1,y  ]
                    coord_up    = [x  ,y-1]
                    coord_down  = [x  ,y+1]

                    for [x_check,y_check] in [coord_down,coord_up,coord_left,coord_right]:
                    
                        # ignore if out of bounds
                        if len(paths[0]) -1 < x_check or x_check <0 or len(paths) -1 < y_check or y_check < 0 :
                            continue
                        
                        #ignore if already int
                        if type(paths[y_check][x_check]["dist_from_E"]) == int and paths[y_check][x_check]["dist_from_E"] < current_square["dist_from_E"]+1:
                            continue

                        next_square = paths[y_check][x_check]
                        current_altitude = altitude.index(current_square["altitude"])
                        next_altitude    = altitude.index(next_square["altitude"])

                        if next_altitude + 2 > current_altitude :
                            paths[y_check][x_check]["dist_from_E"] = current_square["dist_from_E"]+1
        
        i += 1
    
    elements = []
    for row in paths:
        for square in row:
            if type(square["dist_from_E"]) == int and square["altitude"] == 'a':
                elements.append(square["dist_from_E"])

    return min(elements)

def main():
    """
    main function
    """ 
    testfile  = "tests.txt"
    inputfile = "inputs.txt"

    print("Advent-Of-Code 2022 - Day12")

    print(f'"{testfile}": Answer to Part 1 = {Part1(testfile)}')
    print(f'"{testfile}": Answer to Part 2 = {Part2(testfile)}\n')

    print(f'"{inputfile}": Answer to Part 1 = {Part1(inputfile)}')
    print(f'"{inputfile}": Answer to Part 2 = {Part2(inputfile)}')

if __name__ == "__main__":
    main()

File: AdventOfCode2022/Go/Day12/Day12.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
	"slices"
)

type Coord struct {
	x        int
	y        int
	altitude int
	distToE  int
}

func Parse(filename string) ([][]Coord, Coord, Coord) {
	file, _ := os.Open(filename)
	fscanner := bufio.NewScanner(file)

	startCoord := Coord{}
	endCoord := Coord{}
	dataMatrix := [][]Coord{}

	y := 0
	for fscanner.Scan() {
		line := fscanner.Text()
		dataMatrixLine := []Coord{}
		for x, char := range line {
			if char == 'S' {
				startCoord = Coord{x: x, y: y, altitude: 0, distToE: 999}
				dataMatrixLine = append(dataMatrixLine, startCoord)
				continue
			}
			if char == 'E' {
				endCoord = Coord{x: x, y: y, altitude: 25, distToE: 0}
				dataMatrixLine = append(dataMatrixLine, endCoord)
				continue
			}
			altitude := strings.Index("abcdefghijklmnopqrstuvwxyz", string(char))
			coord := Coord{x: x, y: y, altitude: altitude, distToE: 999}
			dataMatrixLine = append(dataMatrixLine, coord)
		}
		dataMatrix = append(dataMatrix, dataMatrixLine)
		y++
	}
	return dataMatrix, startCoord, endCoord
}

func Solve(dataMatrix [][]Coord, startCoord Coord, endCoord Coord) int {
	finalDistance := 0
	for dataMatrix[startCoord.y][startCoord.x].distToE == 999 && finalDistance < 999 {

		for y := 0; y < len(dataMatrix); y++ {

			for x := 0; x < len(dataMatrix[0]); x++ {

				if dataMatrix[y][x].distToE == finalDistance {

					leftCoord := Coord{x: x - 1, y: y}
					rightCoord := Coord{x: x + 1, y: y}
					upCoord := Coord{x: x, y: y - 1}
					downCoord := Coord{x: x, y: y + 1}

					for _, coordPos := range []Coord{leftCoord, rightCoord, upCoord, downCoord} {

						if coordPos.x < 0 || coordPos.x >= len(dataMatrix[0]) || coordPos.y < 0 || coordPos.y >= len(dataMatrix) {
							continue
						}

						coord := dataMatrix[coordPos.y][coordPos.x]

						if coord.distToE != 999 && coord.distToE < dataMatrix[y][x].distToE+1 {
							continue
						}

						if coord.altitude + 2 > dataMatrix[y][x].altitude {
							dataMatrix[coord.y][coord.x].distToE = dataMatrix[y][x].distToE + 1
						}
					}
				}
			}
		}
		finalDistance++
	}
	return finalDistance
}

func Part1(filename string) int {
	dataMatrix, startCoord, endCoord := Parse(filename)
	return Solve(dataMatrix, startCoord, endCoord)
}

func Part2(filename string) int {
	pathLengths := []int{}
	dataMatrix, startCoord, endCoord := Parse(filename)
	Solve(dataMatrix, startCoord, endCoord)
	for _, line := range dataMatrix {
		for _, coord := range line {
			if coord.altitude == 0 {
				pathLengths = append(pathLengths, coord.distToE)
			}
		}
	}
	return slices.Min(pathLengths)
}

func main() {
	fmt.Println("Advent-Of-Code 2022 - Day12")
	fmt.Printf("Tests : Answer to Part 1 = %v\n", Part1("tests.txt"))
	fmt.Printf("Inputs: Answer to Part 1 = %v\n", Part1("inputs.txt"))
	fmt.Printf("Tests : Answer to Part 2 = %v\n", Part2("tests.txt"))
	fmt.Printf("Inputs: Answer to Part 2 = %v\n", Part2("inputs.txt"))
}
